src/caffeine/caffeine.c

```diff
@@ -22,126 +21,40 @@ static gex_Rank_t rank, size;
   const int double_Complex_workaround =4100;
 #endif
 
-
-#define HANDLE_SIZE(corank) (sizeof(intmax_t) + (sizeof(intmax_t) * (corank) * 2) + sizeof(final_func_ptr) + sizeof(size_t) + sizeof(int64_t) + sizeof(char*))
-
-// macros that access a field of the coarray handle (CAH), an internal coarray metadata object
-
-// since the size of the handle is dependent on the value of corank, the first field in the handle, each accessor
-// dereferences corank and uses its value to correctly calculate the memory locations of the other fields
-#define CAH_CORANK(coarray_handle) *((intmax_t*)(coarray_handle))
-#define CAH_LCOBOUNDS(coarray_handle) (intmax_t*)((char*)(coarray_handle) + sizeof(intmax_t))
-#define CAH_UCOBOUNDS(coarray_handle) (intmax_t*)((char*)(coarray_handle) + sizeof(intmax_t) + (sizeof(intmax_t) * *(intmax_t*)(coarray_handle)))
-#define CAH_FINAL_FUNC(coarray_handle) *(final_func_ptr*)((char*)(coarray_handle) + sizeof(intmax_t) + (sizeof(intmax_t) * *(intmax_t*)(coarray_handle) * 2))
-#define CAH_OBJECT_SIZE(coarray_handle) *(size_t*)((char*)(coarray_handle) + sizeof(intmax_t) + (sizeof(intmax_t) * *(intmax_t*)(coarray_handle) * 2) + sizeof(final_func_ptr))
-#define CAH_ALLOCATOR(coarray_handle) *(int64_t*)((char*)(coarray_handle) + sizeof(intmax_t) + (sizeof(intmax_t) * *(intmax_t*)(coarray_handle) * 2) + sizeof(final_func_ptr) + sizeof(size_t))
-
-#define CAH_OBJECT_BASE_ADDR(coarray_handle) *(char**)((char*)(coarray_handle) + sizeof(intmax_t) + (sizeof(intmax_t) * *(intmax_t*)(coarray_handle) * 2) + sizeof(final_func_ptr) + sizeof(size_t) + sizeof(int64_t))
-
-#define PTR_DIFFERENCE_BYTES(first_element_addr, fortran_object_base_addr) ((char*)(first_element_addr) - (char*)(fortran_object_base_addr))
-
-static void setupCoarrayHandle(void* handle_mem, int corank, intmax_t* lcobounds, intmax_t* ucobounds, final_func_ptr final_func, size_t sz, int64_t allocator, void* object_base_addr_ptr);
-
-void caf_caffeinate(int argc, char *argv[])
+void caf_caffeinate(mspace* symmetric_heap, gex_TM_t** initial_team)
 {
-  GASNET_SAFE(gex_Client_Init(&myclient, &myep, &myteam, "caffeine", &argc, &argv, 0));
+  *initial_team = malloc(sizeof(gex_TM_t));
```

Comment:
`gex_TM_t` is a scalar type (a pointer to opaque data), and acts as a trivially-copyable handle.
I would prefer we pass it around by value instead of adding this extra level of indirection on `gex_TM_t`.

If we need an interoperable opaque type, we can safely reinterpret cast a `gex_TM_t` to a `c_intptr_t` or `c_ptr` in the same way we are currently doing for `mspace`.

Reply:
It does seem to work, but just to be sure we understand whether it's supposed to... The reason it works for the `mspace` is because `mspace` is defined as `typedef void* mspace`. Is `gex_TM_t` defined somewhat similarly? I.e. it's guaranteed to have the same size as a `void*`?

Reply:
It works for the same reason: `gex_TM_t` is implemented as a C pointer type (a pointer to an incomplete type, similar to `void *`, and guaranteed by C to have the same size).

`gex_TM_t` is not technically guaranteed by specification to be a pointer type, but that's the current implementation and it's very unlikely to ever change. Our UPC++ implementation is also currently relying on this property, so I have no hesitation recommending the same here.